pkg/frontend/querymiddleware/blocker_test.go

```diff
@@ -0,0 +1,130 @@
+		},
+		{
+			name: "blocks single line query non regex pattern",
+			limits: mockLimits{
+				blockedQueries: []*validation.BlockedQuery{
+					{Pattern: "rate(metric_counter[5m])", Regex: false},
+				},
+			},
+			expectedErr: assert.Error,
+			request: Request(&PrometheusRangeQueryRequest{
+				Query: "rate(metric_counter[5m])",
+			}),
+		},
+		{
+			name: "not blocks single line query non regex pattern",
+			limits: mockLimits{
+				blockedQueries: []*validation.BlockedQuery{
+					{Pattern: "rate(metric_counter[5m])", Regex: false},
+				},
+			},
+			expectedErr: assert.NoError,
+			request: Request(&PrometheusRangeQueryRequest{
+				Query: "rate(metric_counter[15m])",
+			}),
+		},
+		{
+			name: "blocks multiple line query non regex pattern",
+			limits: mockLimits{
+				blockedQueries: []*validation.BlockedQuery{
+					{Pattern: `rate(metric_counter[5m])/
+rate(other_counter[5m])`, Regex: false},
+				},
+			},
+			expectedErr: assert.Error,
+			request: Request(&PrometheusRangeQueryRequest{
+				Query: `rate(metric_counter[5m])/
+rate(other_counter[5m])`,
+			}),
+		},
+		{
+			name: "not blocks multiple line query non regex pattern",
+			limits: mockLimits{
+				blockedQueries: []*validation.BlockedQuery{
+					{Pattern: "rate(metric_counter[5m])", Regex: false},
+				},
+			},
+			expectedErr: assert.NoError,
+			request: Request(&PrometheusRangeQueryRequest{
+				Query: `rate(metric_counter[15m])/
+rate(other_counter[15m])`,
+			}),
+		},
+		{
+			name: "blocks single line query regex pattern",
+			limits: mockLimits{
+				blockedQueries: []*validation.BlockedQuery{
+					{Pattern: ".*metric_counter.*", Regex: true},
+				},
+			},
+			expectedErr: assert.Error,
+			request: Request(&PrometheusRangeQueryRequest{
+				Query: "rate(metric_counter[5m])",
+			}),
+		},
+		{
+			name: "blocks multiple line query regex pattern",
+			limits: mockLimits{
+				blockedQueries: []*validation.BlockedQuery{
+					{Pattern: ".*metric_counter.*", Regex: true},
+				},
+			},
+			expectedErr: assert.Error,
+			request: Request(&PrometheusRangeQueryRequest{
+				Query: `rate(other_counter[15m])/
+rate(metric_counter[15m])`,
+			}),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			reg := prometheus.NewPedanticRegistry()
+			logger := log.NewNopLogger()
+			mw := newQueryBlockerMiddleware(tt.limits, logger, reg)
+			_, err := mw.Wrap(mockHandlerWith(nil, nil)).Do(user.InjectOrgID(context.Background(), "test"), tt.request)
+			tt.expectedErr(t, err)
+			if err != nil {
+				assert.Contains(t, err.Error(), globalerror.QueryBlocked)
+				assert.NoError(t, promtest.GatherAndCompare(reg, strings.NewReader(`
+                        # HELP cortex_query_frontend_blocked_queries_total The total number of requests that were blocked due to queries blocking policy.
+                        # TYPE cortex_query_frontend_blocked_queries_total counter
+						cortex_query_frontend_blocked_queries_total{user="test"} 1
+					`),
```

Comment:
my thinking was that what we want to make sure is that the query doesn't proceed further in the handling chain. We can assert on the metric, but that assumes that there is no bug in the metric counting logic or the handler chain. The request can still proceed while the metric is incremented for example.

It also makes the test somewhat more brittle. If we change the metric which records rejected queries (e.g. add a new label, rename a label), then the test for rejecting queries fails even though the query was properly rejected.

Instead if we assert that the blocking middleware doesn't invoke the next middleware with the query, then we can be more confident that the query doesn't actually run. Instead of injecting a `mockHandlerWith` you can inject a `HandlerFunc` which calls `t.Error()` if it is invoked for a query that should have been blocked.

I agree having 10 different ways of doing this doesn't help 😞 I think we assert on metrics whenever it's difficult to assert on the actions of the real code. For example, the compactor tests use metrics checks quite a lot. For example, some of the tests want to make sure that there is a background cleanup job that runs periodically. If there is no work for the job to do, it's difficult to assert that it has ran. Another reason may be because the test setup is different each time, so it's difficult to know e.g. which block had to marked for deletion and assert on that. Instead the compactor asserts on metrics that _some_ block was marked for deletion.

Reply:
Thank you for explaining. I've tried to implement something based on your recommendation.